### ISSUE TITLE
Adding and renaming child SAs makes SPD go out of sync after a reload 

### ISSUE THREAD

**Issue by toreanderson** [strongswan/strongswan#1046]
**System (please complete the following information):**
 - OS: Rocky Linux release 8.5 (Green Obsidian)
 - Kernel version (if applicable): 4.18.0-348.20.1.el8_5.x86_64
 - strongSwan version(s): strongswan-5.9.5-2.el8.x86_64.rpm (from Fedora EPEL), strongswan-5.9.6 (locally built from source)
 - Tested/confirmed with the latest version: yes

**Describe the bug**
When renaming a child SA and adding a new child SA using the same name that the renamed SA previously had, strongSwan unexpectedly removes the SPD entries for the renamed SA.

**To Reproduce**
1. Start strongSwan: `ipsec start --nofork`
2. Create the file `$PREFIX/etc/swanctl/conf.d/test.conf` with the following contents:
```
connections {
  test {
    remote_addrs=1.2.3.4
    children {
      child1 {
#        start_action=trap
#        local_ts=100.100.100.0/25
#        remote_ts=100.100.100.128/25
#      }
#      child2 {
        start_action=trap
        local_ts=200.200.200.0/25
        remote_ts=200.200.200.128/25
      }
    }
  }
}
```
4. Load the `test` connection: `swanctl -q`
5. Verify that the `test` connection is correctly loaded in strongSwan:
```
$ swanctl -L
test: IKEv1/2, no reauthentication, rekeying every 14400s
  local:  %any
  remote: 1.2.3.4
  local unspecified authentication:
  remote unspecified authentication:
  child1: TUNNEL, rekeying every 3600s
    local:  200.200.200.0/25
    remote: 200.200.200.128/25
```
6. Verify that the the expected entries for `child1` exists in the SPD:
```
$ ip xfrm policy
src 200.200.200.0/25 dst 200.200.200.128/25 
	dir out priority 374400 ptype main 
	tmpl src 10.20.211.106 dst 1.2.3.4
		proto esp reqid 1 mode tunnel
src 200.200.200.128/25 dst 200.200.200.0/25 
	dir fwd priority 374400 ptype main 
	tmpl src 1.2.3.4 dst 10.20.211.106
		proto esp reqid 1 mode tunnel
src 200.200.200.128/25 dst 200.200.200.0/25 
	dir in priority 374400 ptype main 
	tmpl src 1.2.3.4 dst 10.20.211.106
		proto esp reqid 1 mode tunnel
(…entries for 0.0.0.0/0 and ::/0 omitted…)
```
7. Rename `child1` to `child2`, adding a new `child1` by uncommenting lines in `test.conf`: `sed -i 's/#//' $PREFIX/etc/swanctl/conf.d/test.conf`
8. Reload the updated `test` connection: `swanctl -q`
9. Verify that the updated `test` connection is correctly loaded in strongSwan:
```
$ swanctl -L
test: IKEv1/2, no reauthentication, rekeying every 14400s
  local:  %any
  remote: 1.2.3.4
  local unspecified authentication:
  remote unspecified authentication:
  child1: TUNNEL, rekeying every 3600s
    local:  100.100.100.0/25
    remote: 100.100.100.128/25
  child2: TUNNEL, rekeying every 3600s
    local:  200.200.200.0/25
    remote: 200.200.200.128/25
```
10. Observe that the SPD do not contain any entries for `child2`: 
```
$ ip xfrm policy
src 100.100.100.0/25 dst 100.100.100.128/25 
	dir out priority 374400 ptype main 
	tmpl src 10.20.211.106 dst 1.2.3.4
		proto esp reqid 2 mode tunnel
src 100.100.100.128/25 dst 100.100.100.0/25 
	dir fwd priority 374400 ptype main 
	tmpl src 1.2.3.4 dst 10.20.211.106
		proto esp reqid 2 mode tunnel
src 100.100.100.128/25 dst 100.100.100.0/25 
	dir in priority 374400 ptype main 
	tmpl src 1.2.3.4 dst 10.20.211.106
		proto esp reqid 2 mode tunnel
(…entries for 0.0.0.0/0 and ::/0 omitted…)
```

**Expected behavior**
The SPD shown in step 10 should have contained entries for `child2` (the same entries displayed in step 6) .

**Logs/Backtraces**

Terminal output from step 1 during reproducing the issue:

```
$ ipsec start --nofork
Starting strongSwan 5.9.6 IPsec [starter]...
00[DMN] Starting IKE charon daemon (strongSwan 5.9.6, Linux 4.18.0-348.20.1.el8_5.x86_64, x86_64)
00[CFG] loading ca certificates from '/home/rlnc-user/local/strongswan-5.9.6/etc/ipsec.d/cacerts'
00[CFG] loading aa certificates from '/home/rlnc-user/local/strongswan-5.9.6/etc/ipsec.d/aacerts'
00[CFG] loading ocsp signer certificates from '/home/rlnc-user/local/strongswan-5.9.6/etc/ipsec.d/ocspcerts'
00[CFG] loading attribute certificates from '/home/rlnc-user/local/strongswan-5.9.6/etc/ipsec.d/acerts'
00[CFG] loading crls from '/home/rlnc-user/local/strongswan-5.9.6/etc/ipsec.d/crls'
00[CFG] loading secrets from '/home/rlnc-user/local/strongswan-5.9.6/etc/ipsec.secrets'
00[LIB] loaded plugins: charon aes des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem pkcs8 fips-prf gmp curve25519 xcbc cmac hmac kdf drbg attr kernel-netlink resolve socket-default stroke vici updown xauth-generic counters
00[JOB] spawning 16 worker threads
charon (134896) started after 40 ms
10[CFG] added vici connection: test
10[CFG] installing 'child1'
07[CFG] updated vici connection: test
07[CFG] installing 'child1'
07[CFG] updating already routed CHILD_SA 'child1'
```

**Additional context**
The issue causes a outage for production traffic matching the missing SPD entry.

**Comment by tobiasbrunner** [comment]
Yeah, renames are tricky because several things in the IKE daemon are name-based (one example are trap policies, but connections in general are manged by name). So I'd recommend to avoid renaming child configs (in particular reusing already existing names) whenever possible.

While config objects are generally global, immutable and refcounted objects, child configs in peer config objects can actually be replaced. However, when they are replaced, their name is ignored and equal config objects are just replaced so that the new name can be used e.g. to initiate the config via `swanctl --initiate`. But we don't consider them added/removed configs, so existing connections with `start_action=start` are not torn down and restarted due to a rename (before 40ed81244296f708ac6d4cf56d57cf08ca617d66 equal configs were not even replaced, so their name didn't change).

So a rename currently doesn't affect established connections or installed trap policies, where a reference to the original child config is kept (the trap manager even keeps a copy of the name). You see that in the log where the new `child1` config updates the existing trap entry after a name-based lookup, but the renamed `child2` is not installed because it's not considered a new config.

I'm not yet sure how to fix this. One option could be to make the name of existing child configs mutable, but that might have even more unknown/unintended side-effects due to the multi-threaded nature of the daemon (and e.g. copies of names like the one kept by the trap manager). Another option could be to treat renames similar to an actual config replacement but mark them somehow so we don't tear down connections but can still update/replace trap policies.

**Comment by toreanderson** [comment]
For what it is worth, the child SA names are irrelevant for my use case. They are auto generated by our provisioning system, which expands two lists of local and remote subnets to individual child SAs that cover all possible combinations.

The child SA names generated is simply a counter that increments which each generated child SA. So when a subnet is removed or added to the subnet lists, chances are the config file re-generated, the configured names will shuffle around.

So one possible way to approach the problem would be to allow for child SAs to not have an explicit name in the config file. Instead, strongSwan could assign a serial number "name" dynamically - similar to what already happens to established IKE and child SAs.

Another solution would be to make the automation system generate child SA names that are an md5sum of `${local_ts}${remote_ts}` or something along those lines.

**Comment by tobiasbrunner** [comment]
> They are auto generated by our provisioning system, which expands two lists of local and remote subnets to individual child SAs that cover all possible combinations.

Is that for some legacy implementation on the other end that requires separate CHILD_SAs for this?

> So one possible way to approach the problem would be to allow for child SAs to not have an explicit name in the config file. Instead, strongSwan could assign a serial number "name" dynamically - similar to what already happens to established IKE and child SAs.

That's not possible with our config file syntax or the vici protocol.

> Another solution would be to make the automation system generate child SA names that are an md5sum of `${local_ts}${remote_ts}` or something along those lines.

Yes, you can definitely do that. I know one user that does this (they hash all child settings, not only traffic selectors, to get unique names).

**Comment by toreanderson** [comment]
We do indeed frequently encounter peers that do not even support IKEv2, or if they do, do not support multi-subnet child SAs. (We don't control the remote end, our customers/partners/peers do.)

I will probably go for the hashed names workaround.

**Comment by toreanderson** [comment]
When implementing the hashed SA names workaround, I noticed another issue.

When renaming all the child SAs in the swanctl config files, the old child SA names persisted in the output from `swanctl -l`, even after a scheduled rekeying of those child SAs and even after a complete teardown on the IKE and all child SAs (`swanctl --terminate --ike ${conn}`). In the latter case, when the IKE SA is recreated with its children (due to correct and functioning trap policies), those children keep on using the old serial-number based names, even though the output from `swanctl -L` use the new hashed names.

Note that all the child SAs were renamed to new unique names, there is no re-use of old names with new traffic selectors (as was the case in the issue initially described). That is, the renames done were of the form `child1 => d8d0e3704b6954bb7f638a760c15dc2c`, `child2 => 6a5bc1cc5f80a48b540bc09d082b5855` and so on.

In order to make `swanctl -l` display new names, the following procedure works:

```
mv /etc/strongswan/swanctl/conf.d/${conn}.conf /tmp
swanctl -q
swanctl --terminate --ike ${conn}
mv /tmp/${conn}.conf /etc/strongswan/swanctl/conf.d/
swanctl -q
```

No big deal for me, but I thought I should mention it as I found the behaviour unexpected - I had expected that the new names would start being used at least after a `swanctl -q && swanctl --terminate --ike ${conn}` followed by the tunnel being established from scratch.

**Comment by tobiasbrunner** [comment]
Yes, established connections are not (or only partially e.g. for child configs as responder) affected by config changes (not only in regards to renames). And as discussed before, the installed trap policies are currently not renamed.

**Comment by toreanderson** [comment]
To be clear, that established connections did not get renamed was not something I found unexpected.

What I *did* find unexpected is that after I manually tore down the entire connection (which was established before the renaming for Child SAs) using `swanctl --terminate --ike $conn` (which was established prior to the renames), the new connection that was re-established *after* the renames had taken place kept on using the old names.

**Comment by tobiasbrunner** [comment]
> the new connection that was re-established _after_ the renames had taken place kept on using the old names.

Again, that's because the trap policies are not updated/renamed. They still use the config objects with the old names.
